builtin/game/misc.lua

```diff
@@ -4,74 +4,34 @@
 -- Misc. API functions
 --
 
-local timers = {}
-local mintime
-local function update_timers(delay)
-	mintime = false
-	local sub = 0
-	for index = 1, #timers do
-		index = index - sub
-		local timer = timers[index]
-		timer.time = timer.time - delay
-		if timer.time <= 0 then
-			core.set_last_run_mod(timer.mod_origin)
-			timer.func(unpack(timer.args or {}))
-			table.remove(timers, index)
-			sub = sub + 1
-		elseif mintime then
-			mintime = math.min(mintime, timer.time)
-		else
-			mintime = timer.time
-		end
-	end
-end
+local _jobs = {}
+local _time = 0.0
 
-local timers_to_add
-local function add_timers()
-	for _, timer in ipairs(timers_to_add) do
-		table.insert(timers, timer)
-	end
-	timers_to_add = false
-end
-
-local delay = 0
 core.register_globalstep(function(dtime)
-	if not mintime then
-		-- abort if no timers are running
+	_time = _time + dtime
+
+	if #_jobs < 1 then
 		return
 	end
-	if timers_to_add then
-		add_timers()
-	end
-	delay = delay + dtime
-	if delay < mintime then
-		return
+
+	for i = #_jobs, 1, -1 do
+		if _time >= _jobs[i].expire then
+			core.set_last_run_mod(_jobs[i].mod_origin)
+			_jobs[i].func(unpack(_jobs[i].arg))
+			table.remove(_jobs, i)
+		end
 	end
-	update_timers(delay)
-	delay = 0
 end)
 
 function core.after(time, func, ...)
 	assert(tonumber(time) and type(func) == "function",
 			"Invalid core.after invocation")
-	if not mintime then
-		mintime = time
-		timers_to_add = {{
-			time   = time+delay,
-			func   = func,
-			args   = {...},
-			mod_origin = core.get_last_run_mod(),
-		}}
-		return
-	end
-	mintime = math.min(mintime, time)
-	timers_to_add = timers_to_add or {}
-	timers_to_add[#timers_to_add+1] = {
-		time   = time+delay,
-		func   = func,
-		args   = {...},
-		mod_origin = core.get_last_run_mod(),
-	}
+	table.insert(_jobs, {
+		func = func,
+		expire = _time + time,
+		arg = { ... } or {},
```

Comment:
`{...} or {}` will never evaluate the second part.  `{...}` with no arguments is just `{}`.